### PR TITLE
Add new UV_RUN_FORCE type to uv_run_mode to ensure loop runs even after uv_stop is called

### PR DESCRIPTION
**Pull request by Antruly**
# Add new `UV_RUN_FORCE` type to `uv_run_mode` to ensure loop runs even after `uv_stop` is called

## Description

This PR introduces a new type `UV_RUN_FORCE` to the `uv_run_mode` enum in libuv. This addition aims to address an issue where the loop, after having `uv_stop` called before invoking `uv_run`, fails to run as expected due to the `stop_flag` not being reset appropriately.

### Problem

When `uv_stop` is called before `uv_run`, the loop's `stop_flag` is set. If `uv_run` is called subsequently, it exits immediately without running the loop, and without any error indication, because the `stop_flag` is not cleared.

### Solution

A new `UV_RUN_FORCE` type has been added to the `uv_run_mode` enum. When `UV_RUN_FORCE` is used, the `stop_flag` will be forcibly cleared, ensuring that the loop runs as expected.

Initially, the approach considered was to reset the `stop_flag` unconditionally within `uv_run`. However, this could potentially interfere with existing functionality and logic. To avoid such issues, the new type `UV_RUN_FORCE` was introduced to provide this functionality explicitly.

## Code Changes

The following code changes were made:

```c
// In uv.h or the relevant header file
typedef enum {
  UV_RUN_DEFAULT = 0,
  UV_RUN_ONCE,
  UV_RUN_NOWAIT,
  UV_RUN_FORCE = 1024
} uv_run_mode;

// In core.c or the relevant source file
int uv_run(uv_loop_t *loop, uv_run_mode mode) {
  DWORD timeout;
  int r;
  int can_sleep;
  
  // Add code
  if (mode & UV_RUN_FORCE) {
    if (loop->stop_flag != 0)
      loop->stop_flag = 0;
    mode &= ~UV_RUN_FORCE;
  }
  // Add code end

  r = uv__loop_alive(loop);
  if (!r)
    uv_update_time(loop);

  // Existing code...
}

### CODE REVIEW

**Comment by saghul** [comment]
Not sure I like the approach.

The others are modes, but the new one is a flag.

It's also only implemented in windows and with no tests...

What use case do you have which requires resuming a stopped loop?

**Comment by Antruly** [comment]
> Not sure I like the approach.
> 
> The others are modes, but the new one is a flag.
> 
> It's also only implemented in windows and with no tests...
> 
> What use case do you have which requires resuming a stopped loop?

I understand that my initial implementation might not be the most elegant. My goal is to address the issue where uv_stop might be called inadvertently before the loop starts, ensuring that uv_run can still be executed correctly. The primary purpose of uv_stop is to halt an ongoing uv_loop_t, not to prevent uv_run from starting.

Perhaps my approach is not the best. Another way to handle this could be by adding a function like uv_run_force or adding a parameter to uv_run. This could ensure that the loop runs correctly even if uv_stop was called beforehand.

Alternatively, if adding such a method is not desirable, could we handle the situation where uv_stop is called before uv_run by returning a warning or error to indicate that the loop did not execute correctly?

Thank you for your attention and feedback.

**Comment by saghul** [comment]
> My goal is to address the issue where uv_stop might be called inadvertently before the loop starts, ensuring that uv_run can still be executed correctly. The primary purpose of uv_stop is to halt an ongoing uv_loop_t, not to prevent uv_run from starting.

What is it you are doing which has this (rather odd) case?

**Comment by Antruly** [comment]
> > My goal is to address the issue where uv_stop might be called inadvertently before the loop starts, ensuring that uv_run can still be executed correctly. The primary purpose of uv_stop is to halt an ongoing uv_loop_t, not to prevent uv_run from starting.
> 
> What is it you are doing which has this (rather odd) case?

In my project, I use uv_tcp_t to implement a tcpclient with asynchronous I/O based on libuv. Additionally, it needs to support synchronous functionality, such as the connect method. Because connect and uv_run are not in the same thread, there is a scenario where the user calls the connect method and then immediately calls uv_stop after uv_run. Due to multithreading, uv_run does not execute before uv_stop.

Subsequently, my project has a waitConnectFinish method that waits for the connect to complete, determining success based on the connect callback status. If uv_run does not execute correctly and no error is reported, the connect callback will never be executed. As a result, the user will be stuck in the waitConnectFinish method indefinitely.

I have also considered directly checking the uv_loop_t status, but I feel this approach might not be very user-friendly.

Thank you for your attention and feedback.

**Comment by Antruly** [comment]
I would like to propose a different approach to address the issue. Instead of adding a new enum type, we could add a new method uv_run_force to force the execution of uv_run. This method would look like this:

**Comment by saghul** [comment]
> Because connect and uv_run are not in the same thread, there is a scenario where the user calls the connect method and then immediately calls uv_stop after uv_run. Due to multithreading, uv_run does not execute before uv_stop.

You can't call libuv APIs from different threads. Are you doing that?

**Comment by Antruly** [comment]
> > Because connect and uv_run are not in the same thread, there is a scenario where the user calls the connect method and then immediately calls uv_stop after uv_run. Due to multithreading, uv_run does not execute before uv_stop.
> 
> You can't call libuv APIs from different threads. Are you doing that?

I understand that this might be a multithreading issue. However, uv_run does not raise an exception or return a specific value to indicate this situation. For example, it could return a specific value to signal that uv_run did not execute correctly due to uv_stop being called beforehand.

Thank you for your consideration.

**Comment by saghul** [comment]
Your code is broken and you seem to be trying to fix it in the wrong place, sorry.

That said, we don't have good documentation on this so we might want to change what we do.

@libuv/collaborators thoughts?

**Comment by Antruly** [comment]
> Your code is broken and you seem to be trying to fix it in the wrong place, sorry.
> 
> That said, we don't have good documentation on this so we might want to change what we do.
> 
> @libuv/collaborators thoughts?

Thank you for the community support. Perhaps my approach focused too much on solving the immediate problem without considering the bigger picture. I hope you can address this issue comprehensively.

**Comment by bnoordhuis** [comment]
> That said, we don't have good documentation on this so we might want to change what we do.

@saghul can you summarize what you're asking/proposing here?

**Comment by saghul** [comment]
We currently don't specify what should happen after uv_stop, as in, will uv_run just continue happily ever after?

Currently it doesn't, which might or might be unexpected.

Given how long the current behavior has existed, I'd be inclined to document it as such, that is, the loop will never in again once uv_stop is called.

**Comment by vtjnash** [comment]
It seems a bit awkward to add a new flag for this. In my applications (aka Julia), we just call `loop->stop_flag = 0` just before entering the loop (although doing it just after the uv_run returned would also probably be safe). However calling `uv_stop` from another thread is undefined behavior, so I don't think we should expect to allow that at all.
